pydra/engine/task.py

```diff
@@ -0,0 +1,443 @@
+        if self.audit_flags and flags:
+            if self.messenger_args:
+                send_message(make_message(message), messengers=self.messengers, 
+                             **self.messenger_args)
+            else:              
+                send_message(make_message(message), messengers=self.messengers)
+
+    @property
+    def can_resume(self):
+        """Task can reuse partial results after interruption
+        """
+        return self._can_resume
+
+    @classmethod
+    def help(cls, returnhelp=False):
+        """ Prints class help
+        """
+        helpstr = print_help(cls)
+        if returnhelp:
+            return helpstr
+
+    @property
+    def output_names(self):
+        return [f.name for f in dc.fields(self.output_spec)]
+
+    @property
+    def version(self):
+        return self._version
+    
+    def save_set(self, name, inputs, force=False):
+        if name in self._input_sets and not force:
+            raise KeyError('Key {} already saved. Use force=True to override.')
+        self._input_sets[name] = inputs
+
+    @property
+    def checksum(self):
+        return self.inputs.hash
+
+    @abc.abstractmethod
+    async def _run_interface(self, **kwargs):
+        pass
+
+    def result(self, cache_locations=None):
+        result = load_result(self.checksum, 
+                             ensure_list(cache_locations) + 
+                             ensure_list(self._cache_dir))
+        if result is not None:
+            if 'output' in result:
+                output = self.output_spec(**result['output'])
+            return Result(output=output)
+        return None
+
+    @property
+    def cache_dir(self):
+        return self._cache_dir
+    
+    @cache_dir.setter
+    def cache_dir(self, location):
+        self._cache_dir = Path(location)
+
+    def run(self, cache_locations=None, **kwargs):
+        self.inputs = dc.replace(self.inputs, **kwargs)
+        inputs_hash = self.inputs.hash
+        
+        # Eagerly retrieve cached
+        result = load_result(inputs_hash,
+                             ensure_list(cache_locations) + 
+                             ensure_list(self._cache_dir))
+        if result is not None:
+            return result
+        # start recording provenance
+        aid = gen_uuid()
+        self.audit({"@id": "pydra:{}".format(aid),
+                    "startedAtTime": now()},
+                   AUDIT_PROV)
+          
+        # Not cached        
+        if self._cache_dir is None:
+            self.cache_dir = mkdtemp()
+        odir = self.cache_dir / inputs_hash
+        odir.mkdir(parents=True, exist_ok=True)
+        #check_runtime(self._runtime_requirements)
+        #isolate inputs if files
+        #cwd = os.getcwd()
+        #id = record_provenance(self, env)
+        #resources = start_monitor()
+        result = Result(output=None)
+        try:
+            result.output = self._run_interface(**kwargs)
```

Comment:
Do we still pass a `runtime` object (or `_runtime_requirements`) to `_run_interface`? I'm thinking in terms of the existing `_pre_run_hook` and `_post_run_hook`, but perhaps that's a problem for a later day.